circleci: failed tests: TestStoreMetrics

The following test appears to have failed:

[#19981](https://circleci.com/gh/cockroachdb/cockroach/19981):

```
I160707 16:28:45.046301 kv/txn_coord_sender.go:246  txn coordinator: 0.00 txn/sec, 0.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0/0/0 avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples)
I160707 16:28:45.047445 kv/txn_coord_sender.go:246  txn coordinator: 0.00 txn/sec, 0.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0/0/0 avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples)
I160707 16:28:50.043683 kv/txn_coord_sender.go:246  txn coordinator: 0.00 txn/sec, 100.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0/0/0 avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples)
I160707 16:28:50.046573 kv/txn_coord_sender.go:246  txn coordinator: 0.00 txn/sec, 0.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0/0/0 avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples)
I160707 16:28:50.047762 kv/txn_coord_sender.go:246  txn coordinator: 0.00 txn/sec, 0.00/0.00/0.00/0.00 %cmmt/cmmt1pc/abrt/abnd, 0/0/0 avg/σ/max duration, 0.0/0.0/0 avg/σ/max restarts (0 samples)
panic: test timed out after 2m0s

goroutine 15803 [running]:
panic(0x16a1de0, 0xc8206487c0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
testing.startAlarm.func1()
	/usr/local/go/src/testing/testing.go:725 +0x14b
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x3a

goroutine 1 [chan receive, 1 minutes]:
testing.RunTests(0x1ef5010, 0x26f11e0, 0x123, 0x123, 0xc82021f801)
	/usr/local/go/src/testing/testing.go:583 +0x8d2
testing.(*M).Run(0xc820047ef8, 0xc82026dc98)
	/usr/local/go/src/testing/testing.go:515 +0x81
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc820047ef8)
	/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:54 +0x1de
main.main()
	github.com/cockroachdb/cockroach/storage/_test/_testmain.go:644 +0x114

goroutine 17 [syscall, 1 minutes, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1998 +0x1

goroutine 5 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x2a06be0)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1011 +0x64
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:598 +0x8a

goroutine 6609 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205db7a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6575 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1134 +0x3cf
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205daf80)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6570 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*bookie).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/reservation.go:255 +0x262
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205daea0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6726 [select]:
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc820491560, 0xc82047c7b0, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:141 +0x7e6
google.golang.org/grpc/transport.(*Stream).Read(0xc820143340, 0xc82047c7b0, 0x5, 0x5, 0x5ac521, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc570eabe00, 0xc820143340, 0xc82047c7b0, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fc570eabe00, 0xc820143340, 0xc82047c7b0, 0x5, 0x5, 0x8c3ef8, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
google.golang.org/grpc.(*parser).recvMsg(0xc82047c7a0, 0x22, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/rpc_util.go:216 +0xb9
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc8204915f0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6773 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*RaftTransport).RaftMessage(0xc8205944e0, 0x7fc570e79468, 0xc8203f1430, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_transport.go:149 +0x216
github.com/cockroachdb/cockroach/storage._MultiRaft_RaftMessage_Handler(0x19c4c40, 0xc8205944e0, 0x7fc570eb0710, 0xc820bbf980, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft.pb.go:159 +0xd8
google.golang.org/grpc.(*Server).processStreamingRPC(0xc82057c990, 0x7fc570e74538, 0xc8205474d0, 0xc8204a2620, 0xc8205da5e0, 0x26e3bc0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/server.go:607 +0x47a
google.golang.org/grpc.(*Server).handleStream(0xc82057c990, 0x7fc570e74538, 0xc8205474d0, 0xc8204a2620, 0x0)
	/go/src/google.golang.org/grpc/server.go:691 +0x114e
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc82050d7f0, 0xc82057c990, 0x7fc570e74538, 0xc8205474d0, 0xc8204a2620)
	/go/src/google.golang.org/grpc/server.go:352 +0xa0
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/google.golang.org/grpc/server.go:353 +0x9a

goroutine 6554 [IO wait]:
net.runtime_pollWait(0x7fc570e731a8, 0x72, 0xc820b82000)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8205ddb80, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8205ddb80, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8205ddb20, 0xc820b82000, 0x8000, 0x8000, 0x0, 0x7fc570e92050, 0xc82000e058)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc8201265d0, 0xc820b82000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
--
google.golang.org/grpc/transport.(*http2Client).reader(0xc8201a41e0)
	/go/src/google.golang.org/grpc/transport/http2_client.go:795 +0x109
created by google.golang.org/grpc/transport.newHTTP2Client
	/go/src/google.golang.org/grpc/transport/http2_client.go:174 +0xd21

goroutine 6767 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc820bf50a0, 0xecf1079d4, 0x8886dd6, 0x2a05b20, 0xc8203e00c0, 0xc820bf5030, 0xc820a44d20, 0xc820d37e98)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:163 +0x5b1
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc820a44d20, 0xc8203ac840)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:211 +0x78
github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc8203ac900, 0xc820a09f00)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:258 +0x246
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:213 +0x21e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bf5030, 0xc820959d40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6627 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205db7e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6731 [select]:
google.golang.org/grpc/transport.(*http2Client).controller(0xc820bf1a40)
	/go/src/google.golang.org/grpc/transport/http2_client.go:873 +0x5da
created by google.golang.org/grpc/transport.newHTTP2Client
	/go/src/google.golang.org/grpc/transport/http2_client.go:201 +0x15c2

goroutine 6596 [select, 1 minutes]:
google.golang.org/grpc.NewClientStream.func1(0x7fc570e745b0, 0xc8201a41e0, 0xc820530540, 0xc8204305a0)
	/go/src/google.golang.org/grpc/stream.go:151 +0x258
created by google.golang.org/grpc.NewClientStream
	/go/src/google.golang.org/grpc/stream.go:159 +0xab2

goroutine 6723 [select, 1 minutes]:
google.golang.org/grpc.NewClientStream.func1(0x7fc570e745b0, 0xc8201a41e0, 0xc820143260, 0xc820431540)
	/go/src/google.golang.org/grpc/stream.go:151 +0x258
created by google.golang.org/grpc.NewClientStream
	/go/src/google.golang.org/grpc/stream.go:159 +0xab2

goroutine 6721 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func4()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1044 +0x13d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bf5030, 0xc8209593c0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6598 [select]:
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc8205955c0, 0xc8205db110, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:141 +0x7e6
google.golang.org/grpc/transport.(*Stream).Read(0xc820530540, 0xc8205db110, 0x5, 0x5, 0xab8f54, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc570eabe00, 0xc820530540, 0xc8205db110, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fc570eabe00, 0xc820530540, 0xc8205db110, 0x5, 0x5, 0xc820538180, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
google.golang.org/grpc.(*parser).recvMsg(0xc8205db100, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/rpc_util.go:216 +0xb9
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc820595680)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6730 [IO wait]:
net.runtime_pollWait(0x7fc570e73568, 0x72, 0xc820db8000)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8200c2ed0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8200c2ed0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8200c2e70, 0xc820db8000, 0x8000, 0x8000, 0x0, 0x7fc570e92050, 0xc82000e058)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc820024718, 0xc820db8000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
--
google.golang.org/grpc/transport.(*http2Client).reader(0xc820bf1a40)
	/go/src/google.golang.org/grpc/transport/http2_client.go:795 +0x109
created by google.golang.org/grpc/transport.newHTTP2Client
	/go/src/google.golang.org/grpc/transport/http2_client.go:174 +0xd21

goroutine 6724 [select, 1 minutes]:
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc820491320, 0xc82047c6d0, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:141 +0x7e6
google.golang.org/grpc/transport.(*Stream).Read(0xc820143260, 0xc82047c6d0, 0x5, 0x5, 0x1c, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc570eabe00, 0xc820143260, 0xc82047c6d0, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fc570eabe00, 0xc820143260, 0xc82047c6d0, 0x5, 0x5, 0x30, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
google.golang.org/grpc.(*parser).recvMsg(0xc82047c6c0, 0x30, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/rpc_util.go:216 +0xb9
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc82047c6e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6765 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bf5030, 0xc820959d00)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6712 [select, 1 minutes]:
google.golang.org/grpc.NewClientStream.func1(0x7fc570e745b0, 0xc820bf13b0, 0xc8201176c0, 0xc8208e7720)
	/go/src/google.golang.org/grpc/stream.go:151 +0x258
created by google.golang.org/grpc.NewClientStream
	/go/src/google.golang.org/grpc/stream.go:159 +0xab2

goroutine 6534 [chan receive, 1 minutes]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc8205a6c00)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:319 +0x3a
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:320 +0xada

goroutine 6678 [select]:
google.golang.org/grpc/transport.(*http2Client).controller(0xc820bf13b0)
	/go/src/google.golang.org/grpc/transport/http2_client.go:873 +0x5da
created by google.golang.org/grpc/transport.newHTTP2Client
	/go/src/google.golang.org/grpc/transport/http2_client.go:201 +0x15c2

goroutine 6776 [select]:
github.com/cockroachdb/cockroach/storage.(*RaftTransport).processQueue(0xc8205944e0, 0xc820bda9c0, 0x100000001, 0x1, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_transport.go:212 +0x7d1
github.com/cockroachdb/cockroach/storage.RaftSender.SendAsync.func1.1()
	/go/src/github.com/cockroachdb/cockroach/storage/raft_transport.go:276 +0x6b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc820a4e540)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6657 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8204c7b20, 0xc82051e4a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6713 [select, 1 minutes]:
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc8204414a0, 0xc8204c8d30, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:141 +0x7e6
google.golang.org/grpc/transport.(*Stream).Read(0xc8201176c0, 0xc8204c8d30, 0x5, 0x5, 0x40, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc570eabe00, 0xc8201176c0, 0xc8204c8d30, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fc570eabe00, 0xc8201176c0, 0xc8204c8d30, 0x5, 0x5, 0xc8203e00c0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
google.golang.org/grpc.(*parser).recvMsg(0xc8204c8d20, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/rpc_util.go:216 +0xb9
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc8204c8d40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6574 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1115 +0x3cf
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205daf60)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6655 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8204c7b20, 0xc82051e460)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6626 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205db7c0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6571 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:2189 +0xa71
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205daee0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6568 [select, 1 minutes]:
github.com/cockroachdb/cockroach/gossip.(*Gossip).bootstrap.func1()
	/go/src/github.com/cockroachdb/cockroach/gossip/gossip.go:770 +0x237
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc8205dad20)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6440 [chan receive, 1 minutes]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc8202aa9c0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:319 +0x3a
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:320 +0xada

goroutine 6563 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*StorePool).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store_pool.go:290 +0x23f
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0070, 0xc8205da8e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6599 [select, 1 minutes]:
google.golang.org/grpc/transport.(*recvBufferReader).Read(0xc820569d10, 0xc8205a4310, 0x5, 0x5, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:141 +0x7e6
google.golang.org/grpc/transport.(*Stream).Read(0xc820528460, 0xc8205a4310, 0x5, 0x5, 0x37, 0x0, 0x0)
	/go/src/google.golang.org/grpc/transport/transport.go:294 +0x71
io.ReadAtLeast(0x7fc570eabe00, 0xc820528460, 0xc8205a4310, 0x5, 0x5, 0x5, 0x0, 0x0, 0x0)
	/usr/local/go/src/io/io.go:297 +0xe6
io.ReadFull(0x7fc570eabe00, 0xc820528460, 0xc8205a4310, 0x5, 0x5, 0xc820d0ba18, 0x0, 0x0)
	/usr/local/go/src/io/io.go:315 +0x62
google.golang.org/grpc.(*parser).recvMsg(0xc8205a4300, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/rpc_util.go:216 +0xb9
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc820538340)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6536 [select]:
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).startStats(0xc8205a78c0)
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:203 +0xd21
github.com/cockroachdb/cockroach/kv.(*TxnCoordSender).(github.com/cockroachdb/cockroach/kv.startStats)-fm()
	/go/src/github.com/cockroachdb/cockroach/kv/txn_coord_sender.go:188 +0x20
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0070, 0xc82062ef30)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6735 [select]:
google.golang.org/grpc/transport.(*http2Server).controller(0xc8205474d0)
	/go/src/google.golang.org/grpc/transport/http2_server.go:660 +0x5da
created by google.golang.org/grpc/transport.newHTTP2Server
	/go/src/google.golang.org/grpc/transport/http2_server.go:134 +0x84f

goroutine 6562 [chan receive, 1 minutes]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc8205d1f20)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:319 +0x3a
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:320 +0xada

goroutine 6608 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:376 +0x31b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205db780)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6555 [select, 1 minutes]:
google.golang.org/grpc/transport.(*http2Client).controller(0xc8201a41e0)
	/go/src/google.golang.org/grpc/transport/http2_client.go:873 +0x5da
created by google.golang.org/grpc/transport.newHTTP2Client
	/go/src/google.golang.org/grpc/transport/http2_client.go:201 +0x15c2

goroutine 6559 [IO wait]:
net.runtime_pollWait(0x7fc570e734a8, 0x72, 0xc820b9a000)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8205ddfe0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8205ddfe0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).Read(0xc8205ddf80, 0xc820b9a000, 0x8000, 0x8000, 0x0, 0x7fc570e92050, 0xc82000e058)
	/usr/local/go/src/net/fd_unix.go:250 +0x23a
net.(*conn).Read(0xc8201265e8, 0xc820b9a000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/usr/local/go/src/net/net.go:172 +0xe4
--
google.golang.org/grpc.(*Server).handleRawConn(0xc82057c000, 0x7fc570e73ab8, 0xc8201265e8)
	/go/src/google.golang.org/grpc/server.go:318 +0x4ee
created by google.golang.org/grpc.(*Server).Serve
	/go/src/google.golang.org/grpc/server.go:290 +0x38c

goroutine 6780 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*RaftTransport).RaftMessage(0xc8204d6210, 0x7fc570e79468, 0xc8203f1c70, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_transport.go:149 +0x216
github.com/cockroachdb/cockroach/storage._MultiRaft_RaftMessage_Handler(0x19c4c40, 0xc8204d6210, 0x7fc570eb0710, 0xc820300080, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft.pb.go:159 +0xd8
google.golang.org/grpc.(*Server).processStreamingRPC(0xc82057c000, 0x7fc570e74538, 0xc820612630, 0xc8204a27e0, 0xc8203c40a0, 0x26e3bc0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/server.go:607 +0x47a
google.golang.org/grpc.(*Server).handleStream(0xc82057c000, 0x7fc570e74538, 0xc820612630, 0xc8204a27e0, 0x0)
	/go/src/google.golang.org/grpc/server.go:691 +0x114e
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc82062fa40, 0xc82057c000, 0x7fc570e74538, 0xc820612630, 0xc8204a27e0)
	/go/src/google.golang.org/grpc/server.go:352 +0xa0
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/google.golang.org/grpc/server.go:353 +0x9a

goroutine 6615 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*replicaScanner).waitAndProcess(0xc8204c7ea0, 0xecf1079d4, 0x32a41bb, 0x2a05b20, 0xc8203e00c0, 0xc8204c7b20, 0xc820cb21e0, 0xc820d34e98)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:163 +0x5b1
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1.1(0xc820cb21e0, 0xc8204d8f90)
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:211 +0x78
github.com/cockroachdb/cockroach/storage.(*storeRangeSet).Visit(0xc8204d9290, 0xc8204f85c0)
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:258 +0x246
github.com/cockroachdb/cockroach/storage.(*replicaScanner).scanLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/scanner.go:213 +0x21e
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8204c7b20, 0xc8205a4740)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6687 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*RaftTransport).RaftMessage(0xc8205a9050, 0x7fc570e79468, 0xc82050cde0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft_transport.go:149 +0x216
github.com/cockroachdb/cockroach/storage._MultiRaft_RaftMessage_Handler(0x19c4c40, 0xc8205a9050, 0x7fc570eb0710, 0xc82049f580, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/storage/raft.pb.go:159 +0xd8
google.golang.org/grpc.(*Server).processStreamingRPC(0xc820612120, 0x7fc570e74538, 0xc820546480, 0xc820143180, 0xc8205de7c0, 0x26e3bc0, 0x0, 0x0, 0x0)
	/go/src/google.golang.org/grpc/server.go:607 +0x47a
google.golang.org/grpc.(*Server).handleStream(0xc820612120, 0x7fc570e74538, 0xc820546480, 0xc820143180, 0x0)
	/go/src/google.golang.org/grpc/server.go:691 +0x114e
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc820537ab0, 0xc820612120, 0x7fc570e74538, 0xc820546480, 0xc820143180)
	/go/src/google.golang.org/grpc/server.go:352 +0xa0
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/go/src/google.golang.org/grpc/server.go:353 +0x9a

goroutine 6710 [select]:
github.com/cockroachdb/cockroach/rpc.(*Context).runHeartbeat(0xc8202aa2a0, 0xc820ccf600, 0xc820484900, 0xf, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/rpc/context.go:263 +0x68d
github.com/cockroachdb/cockroach/rpc.(*Context).GRPCDial.func1.1()
	/go/src/github.com/cockroachdb/cockroach/rpc/context.go:177 +0x66
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc820441230)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6573 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func4()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1044 +0x13d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8205205b0, 0xc8205daf40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6535 [select, 1 minutes]:
github.com/cockroachdb/cockroach/storage.(*StorePool).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store_pool.go:290 +0x23f
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0070, 0xc8205dea40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6538 [IO wait, 1 minutes]:
net.runtime_pollWait(0x7fc570e72f68, 0x72, 0x0)
	/usr/local/go/src/runtime/netpoll.go:160 +0x60
net.(*pollDesc).Wait(0xc8205dd6b0, 0x72, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
net.(*pollDesc).WaitRead(0xc8205dd6b0, 0x0, 0x0)
	/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
net.(*netFD).accept(0xc8205dd650, 0x0, 0x7fc570eabae0, 0xc82051eda0)
	/usr/local/go/src/net/fd_unix.go:426 +0x27c
net.(*TCPListener).AcceptTCP(0xc820126568, 0xc820037ea8, 0x0, 0x0)
	/usr/local/go/src/net/tcpsock_posix.go:254 +0x4d
--
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc820bd0000, 0xc8205deda0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:142 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:143 +0x62

goroutine 6777 [select, 1 minutes]:
google.golang.org/grpc.NewClientStream.func1(0x7fc570e745b0, 0xc8201a41e0, 0xc8204a2700, 0xc8203aa320)
	/go/src/google.golang.org/grpc/stream.go:151 +0x258
created by google.golang.org/grpc.NewClientStream
	/go/src/google.golang.org
```
Please assign, take a look and update the issue accordingly.

--- a/docs/RFCS/drain_modes.md
+++ b/docs/RFCS/drain_modes.md
@@ -25,7 +25,7 @@ running to `drain-clients` to both.
 
 In our Stopper usage, we've taken to fairly ruthlessly shutting down service of
 many components once `(*Stopper).Quiesce()` is called. In code, this manifests
-itself through copious use of the `(*Stopper).ShouldDrain()` channel even when
+itself through copious use of the `(*Stopper).ShouldQuiesce()` channel even when
 not running inside of a task (or running long-running operations inside tasks).
 
 This was motivated mainly by endless amounts of test failures around leaked

--- a/server/server.go
+++ b/server/server.go
@@ -144,7 +144,7 @@ func NewServer(ctx Context, stopper *stop.Stopper) (*Server, error) {
 		s.stopper,
 	)
 
-	// A custom RetryOptions is created which uses stopper.ShouldDrain() as
+	// A custom RetryOptions is created which uses stopper.ShouldQuiesce() as
 	// the Closer. This prevents infinite retry loops from occurring during
 	// graceful server shutdown
 	//

--- a/storage/client_metrics_test.go
+++ b/storage/client_metrics_test.go
@@ -17,6 +17,7 @@
 package storage_test
 
 import (
+	"sync"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/internal/client"
@@ -54,56 +55,83 @@ func checkCounter(t *testing.T, s *storage.Store, key string, e int64) {
 	}
 }
 
-func verifyStats(t *testing.T, mtc *multiTestContext, storeIdx int) {
-	// Get the current store at storeIdx.
-	s := mtc.stores[storeIdx]
-	// Stop the store at the given index, while keeping the reference to the
-	// store object. ComputeMVCCStats() still works on a stopped store (it needs
+func verifyStats(t *testing.T, mtc *multiTestContext, storeIdxSlice ...int) {
+	var stores []*storage.Store
+	var wg sync.WaitGroup
+
+	mtc.mu.RLock()
+	numStores := len(mtc.stores)
+	// We need to stop the stores at the given indexes, while keeping the reference to the
+	// store objects. ComputeMVCCStats() still works on a stopped store (it needs
 	// only the engine, which is still open), and the most recent stats are still
 	// available on the stopped store object; however, no further information can
 	// be committed to the store while it is stopped, preventing any races during
 	// verification.
-	mtc.stopStore(storeIdx)
+	for _, storeIdx := range storeIdxSlice {
+		stores = append(stores, mtc.stores[storeIdx])
+	}
+	mtc.mu.RUnlock()
 
-	// Compute real total MVCC statistics from store.
-	realStats, err := s.ComputeMVCCStats()
-	if err != nil {
-		t.Fatal(err)
+	wg.Add(numStores)
+	// We actually stop *all* of the Stores. Stopping only a few is riddled
+	// with deadlocks since operations can span nodes, but stoppers don't
+	// know about this - taking all of them down at the same time is the
+	// only sane way of guaranteeing that nothing interesting happens, at
+	// least when bringing down the nodes jeopardizes majorities.
+	for i := 0; i < numStores; i++ {
+		go func(i int) {
+			defer wg.Done()
+			mtc.stopStore(i)
+		}(i)
+	}
+	wg.Wait()
+
+	for _, s := range stores {
+		fatalf := func(msg string, args ...interface{}) {
+			prefix := s.Ident.String() + ": "
+			t.Fatalf(prefix+msg, args...)
+		}
+		// Compute real total MVCC statistics from store.
+		realStats, err := s.ComputeMVCCStats()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Sanity regression check for bug #4624: ensure intent count is zero.
+		if a := realStats.IntentCount; a != 0 {
+			fatalf("expected intent count to be zero, was %d", a)
+		}
+
+		// Sanity check: LiveBytes is not zero (ensures we don't have
+		// zeroed out structures.)
+		if liveBytes := getGauge(t, s, "livebytes"); liveBytes == 0 {
+			fatalf("expected livebytes to be non-zero, was zero")
+		}
+
+		// Ensure that real MVCC stats match computed stats.
+		checkGauge(t, s, "livebytes", realStats.LiveBytes)
+		checkGauge(t, s, "keybytes", realStats.KeyBytes)
+		checkGauge(t, s, "valbytes", realStats.ValBytes)
+		checkGauge(t, s, "intentbytes", realStats.IntentBytes)
+		checkGauge(t, s, "livecount", realStats.LiveCount)
+		checkGauge(t, s, "keycount", realStats.KeyCount)
+		checkGauge(t, s, "valcount", realStats.ValCount)
+		checkGauge(t, s, "intentcount", realStats.IntentCount)
+		checkGauge(t, s, "sysbytes", realStats.SysBytes)
+		checkGauge(t, s, "syscount", realStats.SysCount)
+		// "Ages" will be different depending on how much time has passed. Even with
+		// a manual clock, this can be an issue in tests. Therefore, we do not
+		// verify them in this test.
+
+		if t.Failed() {
+			fatalf("verifyStats failed, aborting test.")
+		}
 	}
 
-	// Sanity regression check for bug #4624: ensure intent count is zero.
-	if a := realStats.IntentCount; a != 0 {
-		t.Fatalf("Expected intent count to be zero, was %d", a)
+	// Restart all Stores.
+	for i := 0; i < numStores; i++ {
+		mtc.restartStore(i)
 	}
-
-	// Sanity check: LiveBytes is not zero (ensures we don't have
-	// zeroed out structures.)
-	if liveBytes := getGauge(t, s, "livebytes"); liveBytes == 0 {
-		t.Fatal("Expected livebytes to be non-zero, was zero")
-	}
-
-	// Ensure that real MVCC stats match computed stats.
-	checkGauge(t, s, "livebytes", realStats.LiveBytes)
-	checkGauge(t, s, "keybytes", realStats.KeyBytes)
-	checkGauge(t, s, "valbytes", realStats.ValBytes)
-	checkGauge(t, s, "intentbytes", realStats.IntentBytes)
-	checkGauge(t, s, "livecount", realStats.LiveCount)
-	checkGauge(t, s, "keycount", realStats.KeyCount)
-	checkGauge(t, s, "valcount", realStats.ValCount)
-	checkGauge(t, s, "intentcount", realStats.IntentCount)
-	checkGauge(t, s, "sysbytes", realStats.SysBytes)
-	checkGauge(t, s, "syscount", realStats.SysCount)
-	// "Ages" will be different depending on how much time has passed. Even with
-	// a manual clock, this can be an issue in tests. Therefore, we do not
-	// verify them in this test.
-
-	if t.Failed() {
-		t.Log(errors.Errorf("verifyStats failed, aborting test."))
-		t.FailNow()
-	}
-
-	// Restart the store at the provided index.
-	mtc.restartStore(storeIdx)
 }
 
 func verifyRocksDBStats(t *testing.T, s *storage.Store) {
@@ -182,8 +210,7 @@ func TestStoreMetrics(t *testing.T) {
 	mtc.waitForValues(roachpb.Key("z"), []int64{5, 5, 5})
 
 	// Verify all stats on store 0 and 1 after addition.
-	verifyStats(t, mtc, 0)
-	verifyStats(t, mtc, 1)
+	verifyStats(t, mtc, 0, 1)
 
 	// Create a transaction statement that fails, but will add an entry to the
 	// sequence cache. Regression test for #4969.
@@ -211,8 +238,7 @@ func TestStoreMetrics(t *testing.T) {
 	checkCounter(t, mtc.stores[1], "replicas", 1)
 
 	// Verify all stats on store0 and store1 after range is removed.
-	verifyStats(t, mtc, 0)
-	verifyStats(t, mtc, 1)
+	verifyStats(t, mtc, 0, 1)
 
 	verifyRocksDBStats(t, mtc.stores[0])
 	verifyRocksDBStats(t, mtc.stores[1])

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -537,9 +537,57 @@ func (m *multiTestContext) addStore() {
 			),
 		)
 	}
+
+	stopper := stop.NewStopper()
 	if len(m.dbs) <= idx {
 		retryOpts := base.DefaultRetryOptions()
-		retryOpts.Closer = m.clientStopper.ShouldQuiesce()
+		retryOpts.Closer = func() chan struct{} {
+			ch := make(chan struct{})
+			// Feed the channel periodically as long as our "real" stopper
+			// is quiescing. Close it when clientStopper closes (which means
+			// we're done). This is awkward, but necessary. We allow stopping
+			// and restarting stores, but their DistSender needs to respect
+			// that as well (since they may be involved in Store tasks).
+			if m.clientStopper.RunAsyncTask(func() {
+				feed := func() bool { // true when closed (and we're done)
+					select {
+					case ch <- struct{}{}:
+					case <-m.clientStopper.ShouldQuiesce():
+						if ch != nil {
+							close(ch)
+							ch = nil
+						}
+						return true
+					}
+					return false
+				}
+
+				for {
+					var grabbedStopper *stop.Stopper
+					// While the stopper is nil, this store is down.
+					for grabbedStopper != nil {
+						m.mu.RLock()
+						if len(m.stoppers) <= idx {
+							grabbedStopper = m.stoppers[idx]
+						}
+						m.mu.RUnlock()
+						if feed() {
+							return
+						}
+					}
+					select {
+					case <-m.clientStopper.ShouldQuiesce():
+						feed() // to make sure chan is closed
+						return
+					case <-grabbedStopper.ShouldQuiesce():
+						feed()
+					}
+				}
+			}) != nil {
+				close(ch)
+			}
+			return ch
+		}()
 		m.distSenders = append(m.distSenders,
 			kv.NewDistSender(&kv.DistSenderContext{
 				Clock:             m.clock,
@@ -552,7 +600,6 @@ func (m *multiTestContext) addStore() {
 		m.dbs = append(m.dbs, client.NewDB(sender))
 	}
 
-	stopper := stop.NewStopper()
 	ctx := m.makeContext(idx)
 	nodeID := roachpb.NodeID(idx + 1)
 	store := storage.NewStore(ctx, eng, &roachpb.NodeDescriptor{NodeID: nodeID})
@@ -632,10 +679,23 @@ func (m *multiTestContext) gossipNodeDesc(g *gossip.Gossip, nodeID roachpb.NodeI
 // All stopped stores must be restarted before multiTestContext.Stop is called.
 func (m *multiTestContext) stopStore(i int) {
 	m.mu.Lock()
+	// Stopping with multiple stoppers (which are not aware of each other) is
+	// messy.
+	// multiTestContextKVTransport needs a read lock to access its stopper and
+	// it's already in a task, so if we simply grabbed a write lock here while
+	// stopping we could deadlock (see #7678).
+	// So we initiate quiescing under a write lock, and then release the lock
+	// during stopping.
+	stopper := m.stoppers[i]
+	m.stoppers[i] = nil
+	go stopper.Quiesce()
+	<-stopper.ShouldQuiesce()
+	m.mu.Unlock()
+	stopper.Stop()
+
+	m.mu.Lock()
 	defer m.mu.Unlock()
 	m.senders[i].RemoveStore(m.stores[i])
-	m.stoppers[i].Stop()
-	m.stoppers[i] = nil
 	m.stores[i] = nil
 }
 

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -219,6 +219,19 @@ func startMultiTestContext(t *testing.T, numStores int) *multiTestContext {
 }
 
 func (m *multiTestContext) Start(t *testing.T, numStores int) {
+	{
+		// Only the fields we nil out below can be injected into m as it
+		// starts up, so fail early if anything else was set (as we'd likely
+		// override it and the test wouldn't get what it wanted).
+		mCopy := *m
+		mCopy.storeContext = nil
+		mCopy.clocks = nil
+		mCopy.clock = nil
+		mCopy.timeUntilStoreDead = 0
+		if !reflect.DeepEqual(mCopy, multiTestContext{}) {
+			t.Fatalf("illegal fields set in multiTestContext: %+v", &mCopy)
+		}
+	}
 	m.t = t
 	m.reenableTableSplits = config.TestingDisableTableSplits()
 

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -181,12 +181,10 @@ type multiTestContext struct {
 	gossips     []*gossip.Gossip
 	storePools  []*storage.StorePool
 	// We use multiple stoppers so we can restart different parts of the
-	// test individually. clientStopper is for 'db', transportStopper is
-	// for 'transports', and the 'stoppers' slice corresponds to the
-	// 'stores'.
+	// test individually. transportStopper is for 'transports', and the
+	// 'stoppers' slice corresponds to the 'stores'.
 	// TODO(bdarnell): now that there are multiple transports, do we
 	// need transportStopper?
-	clientStopper      *stop.Stopper
 	transportStopper   *stop.Stopper
 	engineStoppers     []*stop.Stopper
 	timeUntilStoreDead time.Duration
@@ -245,9 +243,6 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 	if m.rpcContext == nil {
 		m.rpcContext = rpc.NewContext(&base.Context{Insecure: true}, m.clock, m.transportStopper)
 	}
-	if m.clientStopper == nil {
-		m.clientStopper = stop.NewStopper()
-	}
 
 	for i := 0; i < numStores; i++ {
 		m.addStore()
@@ -270,20 +265,33 @@ func (m *multiTestContext) Stop() {
 	go func() {
 		m.mu.RLock()
 		defer m.mu.RUnlock()
-		stoppers := append([]*stop.Stopper{m.clientStopper}, m.stoppers...)
-		stoppers = append(stoppers, m.transportStopper)
-		// Quiesce all the stoppers so that we can stop all stoppers in unison.
-		for _, s := range stoppers {
-			// Stoppers may be nil if stopStore has been called without restartStore.
-			if s != nil {
-				s.Quiesce()
-			}
+
+		// Quiesce everyone in parallel (before the transport stopper) to avoid
+		// deadlocks.
+		var wg sync.WaitGroup
+		wg.Add(len(m.stoppers))
+		for _, s := range m.stoppers {
+			go func(s *stop.Stopper) {
+				defer wg.Done()
+				// Some Stoppers may be nil if stopStore has been called
+				// without restartStore.
+				if s != nil {
+					// TODO(tschottdorf): seems like it *should* be possible to
+					// call .Stop() directly, but then stressing essentially
+					// any test (TestRaftAfterRemove is a good example) results
+					// in deadlocks where a task can't finish because of
+					// getting stuck in addWriteCommand.
+					s.Quiesce()
+				}
+			}(s)
 		}
-		for _, s := range stoppers {
-			if s != nil {
-				s.Stop()
-			}
+		wg.Wait()
+
+		for _, stopper := range m.stoppers {
+			stopper.Stop()
 		}
+		m.transportStopper.Stop()
+
 		for _, s := range m.engineStoppers {
 			s.Stop()
 		}
@@ -474,6 +482,38 @@ func (m *multiTestContext) makeContext(i int) storage.StoreContext {
 	return ctx
 }
 
+func (m *multiTestContext) populateDB(idx int, stopper *stop.Stopper) {
+	if len(m.dbs) <= idx {
+		m.distSenders = append(m.distSenders, nil)
+		m.dbs = append(m.dbs, nil)
+	}
+	retryOpts := base.DefaultRetryOptions()
+	retryOpts.Closer = stopper.ShouldQuiesce()
+	m.distSenders[idx] = kv.NewDistSender(&kv.DistSenderContext{
+		Clock:             m.clock,
+		RangeDescriptorDB: m,
+		TransportFactory:  m.kvTransportFactory,
+		RPCRetryOptions:   &retryOpts,
+	}, m.gossips[idx])
+	sender := kv.NewTxnCoordSender(m.distSenders[idx], m.clock, false, tracing.NewTracer(),
+		stopper, kv.NewTxnMetrics(metric.NewRegistry()))
+	m.dbs[idx] = client.NewDB(sender)
+}
+
+func (m *multiTestContext) populateStorePool(idx int, stopper *stop.Stopper) {
+	if len(m.storePools) <= idx {
+		m.storePools = append(m.storePools, nil)
+	}
+	m.storePools[idx] = storage.NewStorePool(
+		m.gossips[idx],
+		m.clock,
+		m.rpcContext,
+		/* reservationsEnabled */ false,
+		m.timeUntilStoreDead,
+		stopper,
+	)
+}
+
 // AddStore creates a new store on the same Transport but doesn't create any ranges.
 func (m *multiTestContext) addStore() {
 	m.mu.RLock()
@@ -504,6 +544,9 @@ func (m *multiTestContext) addStore() {
 		m.transports = append(m.transports,
 			storage.NewRaftTransport(m.getNodeIDAddress, m.grpcServers[idx], m.rpcContext))
 	}
+
+	stopper := stop.NewStopper()
+
 	if len(m.gossips) <= idx {
 		// Give this store all previous stores as gossip bootstraps.
 		var resolvers []resolver.Resolver
@@ -525,79 +568,11 @@ func (m *multiTestContext) addStore() {
 		if m.timeUntilStoreDead == 0 {
 			m.timeUntilStoreDead = storage.TestTimeUntilStoreDeadOff
 		}
-		m.storePools = append(
-			m.storePools,
-			storage.NewStorePool(
-				m.gossips[idx],
-				m.clock,
-				m.rpcContext,
-				/* reservationsEnabled */ false,
-				m.timeUntilStoreDead,
-				m.clientStopper,
-			),
-		)
+		m.populateStorePool(idx, stopper)
 	}
 
-	stopper := stop.NewStopper()
 	if len(m.dbs) <= idx {
-		retryOpts := base.DefaultRetryOptions()
-		retryOpts.Closer = func() chan struct{} {
-			ch := make(chan struct{})
-			// Feed the channel periodically as long as our "real" stopper
-			// is quiescing. Close it when clientStopper closes (which means
-			// we're done). This is awkward, but necessary. We allow stopping
-			// and restarting stores, but their DistSender needs to respect
-			// that as well (since they may be involved in Store tasks).
-			if m.clientStopper.RunAsyncTask(func() {
-				feed := func() bool { // true when closed (and we're done)
-					select {
-					case ch <- struct{}{}:
-					case <-m.clientStopper.ShouldQuiesce():
-						if ch != nil {
-							close(ch)
-							ch = nil
-						}
-						return true
-					}
-					return false
-				}
-
-				for {
-					var grabbedStopper *stop.Stopper
-					// While the stopper is nil, this store is down.
-					for grabbedStopper != nil {
-						m.mu.RLock()
-						if len(m.stoppers) <= idx {
-							grabbedStopper = m.stoppers[idx]
-						}
-						m.mu.RUnlock()
-						if feed() {
-							return
-						}
-					}
-					select {
-					case <-m.clientStopper.ShouldQuiesce():
-						feed() // to make sure chan is closed
-						return
-					case <-grabbedStopper.ShouldQuiesce():
-						feed()
-					}
-				}
-			}) != nil {
-				close(ch)
-			}
-			return ch
-		}()
-		m.distSenders = append(m.distSenders,
-			kv.NewDistSender(&kv.DistSenderContext{
-				Clock:             m.clock,
-				RangeDescriptorDB: m,
-				TransportFactory:  m.kvTransportFactory,
-				RPCRetryOptions:   &retryOpts,
-			}, m.gossips[idx]))
-		sender := kv.NewTxnCoordSender(m.distSenders[idx], m.clock, false, tracing.NewTracer(),
-			m.clientStopper, kv.NewTxnMetrics(metric.NewRegistry()))
-		m.dbs = append(m.dbs, client.NewDB(sender))
+		m.populateDB(idx, stopper)
 	}
 
 	ctx := m.makeContext(idx)
@@ -704,6 +679,8 @@ func (m *multiTestContext) restartStore(i int) {
 	m.mu.Lock()
 	defer m.mu.Unlock()
 	m.stoppers[i] = stop.NewStopper()
+	m.populateDB(i, m.stoppers[i])
+	m.populateStorePool(i, m.stoppers[i])
 
 	ctx := m.makeContext(i)
 	m.stores[i] = storage.NewStore(ctx, m.engines[i], &roachpb.NodeDescriptor{NodeID: roachpb.NodeID(i + 1)})

--- a/storage/client_test.go
+++ b/storage/client_test.go
@@ -231,6 +231,17 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 		}
 	}()
 
+	m.stores = make([]*storage.Store, numStores)
+	m.storePools = make([]*storage.StorePool, numStores)
+	m.distSenders = make([]*kv.DistSender, numStores)
+	m.dbs = make([]*client.DB, numStores)
+	m.stoppers = make([]*stop.Stopper, numStores)
+	m.senders = make([]*storage.Stores, numStores)
+	m.idents = make([]roachpb.StoreIdent, numStores)
+	m.grpcServers = make([]*grpc.Server, numStores)
+	m.transports = make([]*storage.RaftTransport, numStores)
+	m.gossips = make([]*gossip.Gossip, numStores)
+
 	if m.manualClock == nil {
 		m.manualClock = hlc.NewManualClock(0)
 	}
@@ -244,8 +255,8 @@ func (m *multiTestContext) Start(t *testing.T, numStores int) {
 		m.rpcContext = rpc.NewContext(&base.Context{Insecure: true}, m.clock, m.transportStopper)
 	}
 
-	for i := 0; i < numStores; i++ {
-		m.addStore()
+	for idx := 0; idx < numStores; idx++ {
+		m.addStore(idx)
 	}
 
 	// Wait for gossip to startup.
@@ -483,10 +494,6 @@ func (m *multiTestContext) makeContext(i int) storage.StoreContext {
 }
 
 func (m *multiTestContext) populateDB(idx int, stopper *stop.Stopper) {
-	if len(m.dbs) <= idx {
-		m.distSenders = append(m.distSenders, nil)
-		m.dbs = append(m.dbs, nil)
-	}
 	retryOpts := base.DefaultRetryOptions()
 	retryOpts.Closer = stopper.ShouldQuiesce()
 	m.distSenders[idx] = kv.NewDistSender(&kv.DistSenderContext{
@@ -501,9 +508,6 @@ func (m *multiTestContext) populateDB(idx int, stopper *stop.Stopper) {
 }
 
 func (m *multiTestContext) populateStorePool(idx int, stopper *stop.Stopper) {
-	if len(m.storePools) <= idx {
-		m.storePools = append(m.storePools, nil)
-	}
 	m.storePools[idx] = storage.NewStorePool(
 		m.gossips[idx],
 		m.clock,
@@ -515,10 +519,7 @@ func (m *multiTestContext) populateStorePool(idx int, stopper *stop.Stopper) {
 }
 
 // AddStore creates a new store on the same Transport but doesn't create any ranges.
-func (m *multiTestContext) addStore() {
-	m.mu.RLock()
-	idx := len(m.stores)
-	m.mu.RUnlock()
+func (m *multiTestContext) addStore(idx int) {
 	var clock *hlc.Clock
 	if len(m.clocks) > idx {
 		clock = m.clocks[idx]
@@ -537,43 +538,32 @@ func (m *multiTestContext) addStore() {
 		m.engines = append(m.engines, eng)
 		needBootstrap = true
 	}
-	if len(m.grpcServers) <= idx {
-		m.grpcServers = append(m.grpcServers, rpc.NewServer(m.rpcContext))
-	}
-	if len(m.transports) <= idx {
-		m.transports = append(m.transports,
-			storage.NewRaftTransport(m.getNodeIDAddress, m.grpcServers[idx], m.rpcContext))
-	}
+	m.grpcServers[idx] = rpc.NewServer(m.rpcContext)
+	m.transports[idx] = storage.NewRaftTransport(m.getNodeIDAddress, m.grpcServers[idx], m.rpcContext)
 
 	stopper := stop.NewStopper()
 
-	if len(m.gossips) <= idx {
-		// Give this store all previous stores as gossip bootstraps.
-		var resolvers []resolver.Resolver
-		func() {
-			m.mu.Lock()
-			defer m.mu.Unlock()
-			for _, addr := range m.nodeIDtoAddr {
-				r, err := resolver.NewResolverFromAddress(addr)
-				if err != nil {
-					m.t.Fatal(err)
-				}
-				resolvers = append(resolvers, r)
+	// Give this store all previous stores as gossip bootstraps.
+	var resolvers []resolver.Resolver
+	func() {
+		m.mu.Lock()
+		defer m.mu.Unlock()
+		for _, addr := range m.nodeIDtoAddr {
+			r, err := resolver.NewResolverFromAddress(addr)
+			if err != nil {
+				m.t.Fatal(err)
 			}
-		}()
-		m.gossips = append(m.gossips, gossip.New(m.rpcContext, resolvers, m.transportStopper))
-		m.gossips[idx].SetNodeID(roachpb.NodeID(idx + 1))
-	}
-	if len(m.storePools) <= idx {
-		if m.timeUntilStoreDead == 0 {
-			m.timeUntilStoreDead = storage.TestTimeUntilStoreDeadOff
+			resolvers = append(resolvers, r)
 		}
-		m.populateStorePool(idx, stopper)
+	}()
+	m.gossips[idx] = gossip.New(m.rpcContext, resolvers, m.transportStopper)
+	m.gossips[idx].SetNodeID(roachpb.NodeID(idx + 1))
+	if m.timeUntilStoreDead == 0 {
+		m.timeUntilStoreDead = storage.TestTimeUntilStoreDeadOff
 	}
 
-	if len(m.dbs) <= idx {
-		m.populateDB(idx, stopper)
-	}
+	m.populateStorePool(idx, stopper)
+	m.populateDB(idx, stopper)
 
 	ctx := m.makeContext(idx)
 	nodeID := roachpb.NodeID(idx + 1)
@@ -618,14 +608,14 @@ func (m *multiTestContext) addStore() {
 	// (these must be populated before the store is started so that
 	// FirstRange() can find the sender)
 	m.mu.Lock()
-	m.stores = append(m.stores, store)
-	m.stoppers = append(m.stoppers, stopper)
+	m.stores[idx] = store
+	m.stoppers[idx] = stopper
 	sender := storage.NewStores(clock)
 	sender.AddStore(store)
-	m.senders = append(m.senders, sender)
+	m.senders[idx] = sender
 	// Save the store identities for later so we can use them in
 	// replication operations even while the store is stopped.
-	m.idents = append(m.idents, store.Ident)
+	m.idents[idx] = store.Ident
 	m.mu.Unlock()
 	if err := store.Start(stopper); err != nil {
 		m.t.Fatal(err)

--- a/storage/replica_range_lease.go
+++ b/storage/replica_range_lease.go
@@ -69,7 +69,7 @@ func (p *pendingLeaseRequest) RequestPending() *roachpb.Lease {
 // opposed to an extension, or acquiring the lease when none is held).
 //
 // Note: Once this function gets a context to be used for cancellation, instead
-// of replica.store.Stopper().ShouldDrain(), care will be needed for cancelling
+// of replica.store.Stopper().ShouldQuiesce(), care will be needed for cancelling
 // the Raft command, similar to replica.addWriteCmd.
 func (p *pendingLeaseRequest) InitOrJoinRequest(
 	replica *Replica,

